### DISCUSSION TITLE
Proposal: with assignment operator

### DISCUSSION THREAD

**Discussion by kofifus** [dotnet/csharplang#6371] [Language Ideas]
Allow a `with=` mutation assignment operator similar to how  the other assignment operators work
```
record EmployeeData(string Name, int Salary);
var daveData = new EmployeeData("Dave", 100000);
```

then Instead of:
```
daveData = daveData with { Salary = 110000 };
```

Allow
```
daveData with= { Salary = 110000 };
```

**Comment by WeihanLi** [comment]
`with` would create a new instance, maybe not so suitable for this case(for this example)

**Comment by kofifus** [comment]
@WeihanLi I think on the contrary it only makes sense because its a new instance, if it was modifying itself there would be no reason to self assign

**Reply by WeihanLi** [reply]
Yeah, you're right. Please ignore me since I think the sample is not completely showing the advantage, it's more helpful when dealing with a complex object. For the same above I think it could be `daveData.Salary = 110000;` that may be more simple.

Another sample may like follows:

``` c#
public record ModelA
{
   public string A { get; set; }
   public string A1 { get; set; }  
   public ModelB B {get; set;}
} 

public record ModelB
{
   public string A { get; set; }
   public string B { get; set; }
} 

var a = new ModelA()
{
  A = "aaa",
  B = new ModelB()
  {
     A = "abb",
     B = "bbb"
  }
};

a with= { B = new ModelB { A="aab" } };
```
